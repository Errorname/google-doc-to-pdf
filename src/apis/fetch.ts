import crossFetch from 'cross-fetch'
import Blob from '../polyfills/Blob'

export const fetch = (token: Function) => (
  url: string,
  body: any = null,
  options: { method?: string; headers?: any; raw?: boolean } = {}
): Promise<Response> => {
  const { method, headers, raw, ...rest } = options

  if (body && !(body instanceof Blob)) {
    body = JSON.stringify(body)
  }

  return crossFetch(url, {
    method: method || body ? 'POST' : 'GET',
    body,
    headers: {
      authorization: `Bearer ${token()}`,
      'Content-Type': 'application/json',
      ...headers
    },
    ...rest
  }).then(r => (raw ? r : r.json()))
}

export const multipart = (
  parts: {
    data: string
    [propName: string]: string
  }[],
  boundary: string
): Blob => {
  const body = []

  parts.map(part => {
    const { data, ...headers } = part
    body.push(
      [
        `--${boundary}`,
        ...Object.entries(headers).map(([key, value]) => `${key}: ${value}`),
        '\n'
      ].join('\n')
    )
    body.push(data)
    body.push('\n\n')
  })

  body.push(`--${boundary}--`)

  return new Blob(body)
}
