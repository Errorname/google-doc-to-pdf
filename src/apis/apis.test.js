import apis from './apis'

jest.mock('./fetch', () => ({
  fetch: token => url => `${url} / (token: ${token})`
}))

const allApis = apis('my-token')

test('drive', () => {
  expect(allApis.drive.create()).toMatchSnapshot()
  expect(allApis.drive.get('abcd123')).toMatchSnapshot()
  expect(allApis.drive.copy('abcd123')).toMatchSnapshot()
  expect(allApis.drive.export('abcd123', 'plain/text')).toMatchSnapshot()
  // expect(allApis.drive.ids(42)).toMatchSnapshot()
})

test('docs', () => {
  expect(allApis.docs.get('abcd123')).toMatchSnapshot()
  expect(allApis.docs.update('abcd123')).toMatchSnapshot()
  // expect(allApis.docs.export('abcd123', 'pdf')).toMatchSnapshot()
})
