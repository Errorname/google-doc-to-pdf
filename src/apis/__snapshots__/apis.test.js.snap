// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docs 1`] = `"https://docs.googleapis.com/v1/documents/abcd123 / (token: my-token)"`;

exports[`docs 2`] = `"https://docs.googleapis.com/v1/documents/abcd123:batchUpdate / (token: my-token)"`;

exports[`drive 1`] = `"https://www.googleapis.com/upload/drive/v3/files?uploadType=multipart&alt=json / (token: my-token)"`;

exports[`drive 2`] = `"https://www.googleapis.com/drive/v3/files/abcd123?fields=parents / (token: my-token)"`;

exports[`drive 3`] = `"https://www.googleapis.com/drive/v3/files/abcd123/copy / (token: my-token)"`;

exports[`drive 4`] = `"https://www.googleapis.com/drive/v3/files/abcd123/export?mimeType=plain/text / (token: my-token)"`;
